### DISCUSSION TITLE
Sharing here an alternative way of UI testing for Avalonia

### DISCUSSION THREAD

**Discussion by alexandrehtrb** [AvaloniaUI/Avalonia#12645] [Ideas]
Hello,

## Update 2025-02-04

I've made a NuGet package for live, on-screen UI tests for Avalonia apps.
The package source code and a working example are available in the repo:

Repo: https://github.com/alexandrehtrb/AlexandreHtrb.AvaloniaUITest
NuGet: https://www.nuget.org/packages/AlexandreHtrb.AvaloniaUITest

--------

I want to share here an idea that I made for UI tests on Avalonia.
There are currently two ways for UI tests on Avalonia: Headless UI testing and Appium.

Although they are good, I believe that an alternative way could be better.
Headless UI testing does not run on a visible app and I really would like to see this kind of testing visually.
Appium seemed a little complicated for having to install drivers and run on a separate process.
I also wanted to run UI tests on Release builds, as close as possible to a production build, for more trustworthiness (especially with new NativeAOT compilation).

Because of this, I made a simple UI test framework. It is far from perfect, but it shows the general idea. I based it on Android Espresso.

**(link above)**

**Comment by grokys** [comment]
I'd be interesting in tracking the progress of any alternative to Appium for UI testing. Appium is pretty unstable on desktop, and since WinAppDriver has been abandoned by MS we're stuck on Appium 1.

**Comment by Mrxx99** [comment]
I recently tested FlaUI (https://github.com/FlaUI/FlaUI) which also worked with Avlonia (only did a very simple test for now). But only runs on Windows.

**Comment by rbarriuso** [comment]
I recently posted a [question on S.O. related to this topic](https://stackoverflow.com/questions/77024752/is-it-possible-to-run-appium-e2e-tests-for-an-avalonia-11-application-on-linux). Does anybody know how to do E2E testing of an Avalonia 11 app on Linux?

**Comment by alexandrehtrb** [comment]
I updated the code and it's a little better now

> I recently posted a [question on S.O. related to this topic](https://stackoverflow.com/questions/77024752/is-it-possible-to-run-appium-e2e-tests-for-an-avalonia-11-application-on-linux). Does anybody know how to do E2E testing of an Avalonia 11 app on Linux?

The code I shared works for all platforms

**Reply by DamianSuess** [reply]
@alexandrehtrb  That's a pretty neat project you got started there. Though it's been a while since the last update, do you have any plans on returning back to it?

Currently, I'm looking for a solutions for adding automated UI testing. Thus far, besides your solution i'm finding:

* https://github.com/VerifyTests/Verify.Avalonia
* https://github.com/IngvarX/Camelot - _last update, Nov 2023_

**Reply by maxkatz6** [reply]
> Currently, I'm looking for a solutions for adding automated UI testing

You are looking too far :)

https://docs.avaloniaui.net/docs/concepts/headless/
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Testing/TestableApp.Headless.XUnit

Camelot project used old version of headless, which required more setup work.

**Reply by DamianSuess** [reply]
> You are looking too far :)

@maxkatz6 you're a good man, thank you!  I'll give it a shot.
Ultimately, we want to automate some QA testing behaviors of the existing GUI built on Prism.Avalonia. I appreciate your quick feedback. That was a late-night response, hope you got your rest last night

**Reply by alexandrehtrb** [reply]
Hi @DamianSuess , the project is pretty much finished, if you are interested in more examples, I use the same framework in [another project of mine](https://github.com/alexandrehtrb/Pororoca/tree/develop/src/Pororoca.Desktop/UITesting).

This UI test framework I shared here is visual, it allows you to see the test running with the UI being updated. It is based on [Android Espresso](https://developer.android.com/training/testing/espresso).

Headless testing, mentioned by @maxkatz6 , runs faster and the tests can run in parallel; however, you can't see the UI being updated, which is something I like in UI testing. This framework is like [Robolectric](https://robolectric.org/) for Android.

**Comment by alexandrehtrb** [comment]
Update: I've made a NuGet package with the code from the example, I updated the first post with the links

**Reply by DamianSuess** [reply]
That is an elegant and simple solution, thank you!

I can't wait to give it a shot this upcoming week and that your package gains traction. 👍
